android/src/main/java/io/ably/lib/push/SharedPreferenceStorage.java

```diff
@@ -0,0 +1,52 @@
+package io.ably.lib.push;
+
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+
+import java.lang.reflect.Field;
+
+public class SharedPreferenceStorage implements Storage{
+
+    private final ActivationContext activationContext;
+
+    public SharedPreferenceStorage(ActivationContext activationContext) {
+        this.activationContext = activationContext;
+    }
+
+    private SharedPreferences sharedPreferences() {
+        return PreferenceManager.getDefaultSharedPreferences(activationContext.getContext());
+    }
+
+    @Override
+    public void putString(String key, String value) {
+        sharedPreferences().edit().putString(key, value).apply();
+    }
+
+    @Override
+    public void putInt(String key, int value) {
+        sharedPreferences().edit().putInt(key, value).apply();
+    }
+
+    @Override
+    public String getString(String key, String defaultValue) {
+        return sharedPreferences().getString(key, defaultValue);
+    }
+
+    @Override
+    public int getInt(String key, int defaultValue) {
+        return sharedPreferences().getInt(key, defaultValue);
+    }
+
+    @Override
+    public void clear(Field[] fields) {
+        SharedPreferences.Editor editor = activationContext.getPreferences().edit();
+        for (Field f : fields) {
+            try {
+                editor.remove((String) f.get(null));
+            } catch (IllegalAccessException e) {
```

Comment:
I was initially confused as to why it is necessary to catch this exception from the [remove](https://developer.android.com/reference/android/content/SharedPreferences.Editor#remove(java.lang.String)) method (it's unchecked and not documented as far as I can see).

But then I looked at the `SharedPreferences.Editor` API and spotted the [clear](https://developer.android.com/reference/android/content/SharedPreferences.Editor#clear()) method. Is there a reason you can't just use that rather than manually iterate fields?

Reply:
We surely can use `clear` method. The reason why I did it this way is because also origin implementation specified which properties should be removed (wasn't removing all of them automatically). I'm not sure if something else is stored there what we don't want to be removed.

Reply:
Fair enough and thanks for explaining. I understand this much better now.